### PR TITLE
fix: clean unsafe assistant content during assembly

### PR DESCRIPTION
**Pull request by Tosko4**
## Summary

Adds an active-context-only cleanup layer before assembled messages are returned to Hermes. Assistant turns that contain only blank structured text or internal thinking/reasoning blocks are dropped from provider replay, while stored rows and DAG content remain unchanged.

Tool-call messages are preserved, then the existing active-context tool-pair sanitizer repairs provider sequencing after cleanup.

## Why

Raw history can safely keep odd provider transcript shapes, but active prompt input still needs to satisfy provider constraints. Blank or thinking-only assistant turns are archive-safe but prompt-hostile.

## Validation

- [x] Focused validation: `pytest -q tests/test_lcm_engine.py::TestAssemblyToolPairGuardrail` -> `16 passed`
- [x] Default validation:
  - [x] `pytest tests/test_lcm_core.py tests/test_lcm_engine.py tests/test_packaging_install.py -q` -> `506 passed`
  - [x] `pytest -q` -> `552 passed`
  - [x] `python -m compileall -q .`
  - [x] `python -m py_compile scripts/import_lossless_claw.py`
  - [x] `bash -n scripts/install.sh scripts/update.sh`
  - [x] `git diff --check`

## Notes

This does not mutate the message store, source mappings, or summary DAG. It only changes the active context returned to the host.

Codex flagged the no-compaction path cursor. The current head resets `_ingest_cursor` to the sanitized active-context length when cleanup removes messages before returning without a compaction pass.

Refs #140

### CODE REVIEW

File: engine.py [hunk 1]
```diff
@@ -731,7 +742,7 @@ def compress(self, messages: List[Dict[str, Any]],
                     compressed,
                     assembly_cap_override=recovery_assembly_cap,
                 )
-            return messages
+            return self._sanitize_active_context_messages(messages)
```
[the remark this answers is not in the document]
Comment (Tosko4) [reply engine.py]: Fixed in cc6b3bf. The no-compaction cleanup path now resets `_ingest_cursor` to the sanitized active-context length when messages are dropped. Added a next-turn persistence regression. Full suite: 552 passed.

**Comment by Tosko4** [comment]
@codex review

**Comment by stephenschoettler** [comment]
Agreed with the latest review. This should block merge until it is fixed or we have evidence it is unrelated.

**Review by stephenschoettler — COMMENTED** [review]
I did a focused maintainer pass on PR #149 at `cc6b3bff4f82`.

Verdict:
- Not ready yet.
- I found one blocking correctness/data-integrity regression in restart/session-rebind ingestion.

Validation:
- Clean temporary checkout: `/tmp/hermes-lcm-pr149.gAQGYi`
- Checked out PR head: `cc6b3bff4f8241fd116390df624b7a2bc9cbd1fc`
- `git diff --check origin/main...HEAD` -> pass
- `ulimit -n 8192; python -m pytest tests/test_lcm_engine.py::TestAssemblyToolPairGuardrail -q` -> `16 passed`
- `ulimit -n 8192; python -m pytest tests/test_lcm_core.py tests/test_lcm_engine.py tests/test_packaging_install.py -q` -> `506 passed`
- `ulimit -n 8192; python -m pytest -q` -> `552 passed`
- `python -m compileall -q .` -> pass
- `python -m py_compile scripts/import_lossless_claw.py` -> pass
- `bash -n scripts/install.sh scripts/update.sh` -> pass
- Custom restart/rebind scratch repro -> exposed blocker: store count went from 4 raw rows to 8 after replaying sanitized active context plus one new user message.

This is a comment review, not an approval.

**Comment by stephenschoettler** [comment]
Please rebase this onto current main and rerun the focused validation. I want to review this on a current branch before deciding whether it is merge-ready.

**Comment by Tosko4** [comment]
@stephenschoettler done, please validate

**Review by stephenschoettler — COMMENTED** [review]
I did a focused maintainer pass on PR #149 at `b2908ea3c9a9`.

Critical:
- engine.py:766-773, engine.py:2011-2062, engine.py:2223-2292 , Restart/session-rebind ingestion still duplicates sanitized active context.
  - The PR correctly stores the original unsafe assistant message, returns a sanitized active context, and resets `_ingest_cursor` for the same live engine.
  - But after a process restart/session rebind, `_ingest_cursor` is gone. Reconciliation tries to match the sanitized active context against the durable store tail. That durable tail still contains the dropped blank assistant row, so the sanitized replay no longer matches any suffix.
  - Result: reconciliation returns cursor 0 and `_ingest_messages()` appends the entire sanitized replay plus the new user message as if it were all new history.
  - Reproduced on current PR head `b2908ea3c9a97bb3752beb04595a4b65ccb8c6a5`:
    - First engine: `sanitized_len=3`, `store_count_1=4`, `cursor_1=3`
    - Rebound second engine with same DB, replayed sanitized context plus one new user message
    - Actual: `store_count_2=8`
    - Expected: `store_count_2=5`
    - Stored contents became:
      `['sys', 'question', '[{"text": "", "type": "text"}]', 'visible answer', 'sys', 'question', 'visible answer', 'new follow-up']`
  - This is a data-integrity regression: raw history gains duplicate replay rows after a normal restart/rebind shape.

Warnings:
- tests/test_lcm_engine.py:7793 covers same-process cursor reset, but not the restart/rebind path. The missing regression is two `LCMEngine` instances sharing the same DB, with `on_session_start()` on both, replaying sanitized active context plus one new turn after the first engine shuts down.

Suggestions:
- Make ingest reconciliation tolerant of active-context-only dropped assistant noise when matching replayed active context against durable store tail, or otherwise persist enough cursor/replay state to distinguish sanitized replay from new delta after rebind.
- Add a regression asserting the restart/rebind case stores only the new turn: original count 4 -> final count 5, not 8.

Looks good:
- The cleanup helper is conservative for active replay:
  - Drops assistant messages only when they have no tool calls and no visible content.
  - Preserves assistant tool-call messages even when content is internal-only.
  - Preserves unknown structured part types instead of risking visible-content loss.
  - Runs the existing tool-pair sanitizer after cleanup, preserving/repairing provider-valid adjacency.
- Same-process no-compaction cursor handling is improved and covered.
- Stored rows are not mutated by the cleanup itself.

Validation:
- Clean temporary checkout: `/tmp/hermes-lcm-pr149-review.4bNU6u`
- Checked out PR head: `b2908ea3c9a97bb3752beb04595a4b65ccb8c6a5`
- Base checked: `origin/main` at `c3626bd75be0b7cd43af03abafb584c97d99cf20`
- `git diff --check origin/main...HEAD` -> pass
- `python -m pytest -q tests/test_lcm_engine.py::TestAssemblyToolPairGuardrail` -> `16 passed, 18 warnings`
- `python -m pytest tests/test_lcm_core.py tests/test_lcm_engine.py tests/test_packaging_install.py -q` -> `524 passed, 40 warnings`
- `python -m pytest -q` -> `571 passed, 40 warnings`
- `python -m compileall -q .` -> pass
- `python -m py_compile scripts/import_lossless_claw.py` -> pass
- `bash -n scripts/install.sh scripts/update.sh` -> pass
- Custom restart/rebind scratch repro -> failed as described above: store count `4 -> 8`

**Comment by stephenschoettler** [comment]
Rechecked latest head `55ad2eb540f5` after the main merge. CI is green, but this is still blocked.

The restart/session-rebind repro still duplicates sanitized active context:

- First engine stores 4 raw rows, including the blank assistant row.
- Returned active context is sanitized to 3 rows.
- Second engine rebinds the same DB/session and ingests that sanitized context plus one new user message.
- Actual durable row count: 8.
- Expected durable row count: 5.
- Reconciliation falls back to cursor 0 and persists the sanitized replay as new history.

This is the same data-integrity issue from the prior review, still present on the latest head. The fix needs restart/rebind reconciliation to tolerate active-context-only dropped assistant rows, or persist enough cursor/replay state to distinguish sanitized replay from new delta after rebind.

Please add the two-engine regression as part of the fix: original raw count 4, after rebind plus one new message final count 5, not 8.

**Review by stephenschoettler — COMMENTED** [review]
Superseded by my latest maintainer comment: https://github.com/stephenschoettler/hermes-lcm/pull/149#issuecomment-4414544293

No additional action beyond that comment.

**Comment by Tosko4** [comment]
@stephenschoettler pushed `66de80a`. I added the two-engine restart/rebind regression you asked for: first engine stores 4 raw rows and returns 3 sanitized active rows, second engine rebinds the same DB/session, then final durable row count is 5 instead of 8.

The fix keeps raw storage lossless. Reconciliation now matches sanitized active replay against a durable-tail view with only active-context-droppable assistant rows removed, so the missing blank assistant replay row no longer makes the whole prefix look like new history.

Local validation passed: focused regression plus guardrail class, core/engine/packaging slice, full pytest, compile checks, import script py_compile, shell syntax checks, and diff check. CI is green on the current head. Could you re-review when you get a chance?

**Comment by Tosko4** [comment]
@codex review

**Comment by stephenschoettler** [comment]
Good question. I want this settled before merge. Please answer here or push the adjustment, whichever is cleaner.

File: engine.py [hunk 2]
```diff
@@ -752,7 +763,14 @@ def compress(self, messages: List[Dict[str, Any]],
                     compressed,
                     assembly_cap_override=recovery_assembly_cap,
                 )
-            return messages
+            sanitized_messages = self._sanitize_active_context_messages(messages)
```
[the remark this answers is not in the document]
Comment (Tosko4) [reply engine.py]: Fixed in 69f54ea. The no-compaction cleanup path now disables synthetic missing-result stubs, so it can drop unsafe assistant-only noise without returning an untracked tool row. Added a restart/rebind regression for a pending assistant tool call: final durable history stays system/user/assistant/user, with no synthetic tool row persisted.

**Comment by Tosko4** [comment]
@stephenschoettler pushed `69f54ea` as a follow-up to the Codex P2.

The no-compaction cleanup path no longer returns synthetic missing-result stubs. It still drops unsafe assistant-only noise, but pending assistant tool calls are left as-is rather than adding an untracked tool row. I added a restart/rebind regression for that shape too, so the durable store stays `system/user/assistant/user` and does not persist a synthetic tool result.

Local validation passed again: guardrail class, core/engine/packaging slice, full pytest, compile checks, import script py_compile, shell syntax checks, and diff check. CI is green on `69f54ea`. Could you re-review this head instead?

**Comment by Tosko4** [comment]
@codex review

File: engine.py [hunk 3]
```diff
@@ -2452,7 +2521,99 @@ def _serialize_messages(self, messages: List[Dict[str, Any]]) -> str:
...
+        provider replay context, and this helper never mutates stored rows,
+        source mappings, or DAG nodes.
+        """
+        cleaned: list[Dict[str, Any]] = []
+        dropped_assistant_messages = 0
+        for msg in messages:
+            if self._should_drop_active_assistant_message(msg):
+                dropped_assistant_messages += 1
+                continue
+            cleaned.append(msg)
```
[the remark this answers is not in the document]
Comment (Tosko4) [reply engine.py:2739]: Fixed in 7299104. Active-context cleanup now strips internal/reasoning parts from mixed assistant content instead of only dropping internal-only turns. Raw stored rows remain unchanged. Added coverage for mixed structured content and string reasoning tags, with the broad suite green.

**Comment by Tosko4** [comment]
@stephenschoettler pushed `7299104` for the latest Codex P1.

Active-context cleanup now strips internal/reasoning content out of mixed assistant messages instead of only dropping internal-only assistant turns. The raw store still keeps the original content unchanged. I added coverage for both structured mixed content and string reasoning tags.

Local validation passed again: guardrail class, core/engine/packaging slice, full pytest, compile checks, import script py_compile, shell syntax checks, and diff check. CI is green on `7299104`. Could you re-review this head?

**Comment by Tosko4** [comment]
@codex review

File: engine.py [hunk 4]
```diff
@@ -1990,6 +2011,52 @@ def _matches_store_tail_suffix(
...
+
+        Raw storage remains lossless. This view is used only to reconcile a
+        restarted process when the host replays sanitized active context that no
+        longer contains assistant messages with no visible content.
+        """
+        return [
+            identity
+            for identity in stored_tail
+            if not self._is_active_context_droppable_identity(identity)
+        ]
```
[the remark this answers is not in the document]
Comment (Tosko4) [reply engine.py]: Fixed in a720617. Restart replay reconciliation now applies the same active-context cleanup identity to retained assistant rows, so stripped mixed content matches the sanitized replay without duplicating durable history. Added a two-engine rebind regression for this exact shape, with the branch rebased onto current main and the broad suite green.

**Comment by Tosko4** [comment]
@stephenschoettler pushed `a720617` on top of current main.

This also addresses the latest Codex P2: restart replay reconciliation now compares against the same active-context cleanup identity for retained assistant rows, not only rows that cleanup removes entirely. The raw store still keeps the original content unchanged.

I added the two-engine rebind regression for stripped mixed assistant content. Local validation passed again, including the guardrail class, core/engine/packaging slice, full pytest, compile checks, import script py_compile, shell syntax checks, and diff check. CI is green on the rebased head.

**Comment by Tosko4** [comment]
@codex review

**Comment by stephenschoettler** [comment]
I agree with Tosko4's review. I am holding merge until this is fixed or there is clear evidence the concern does not apply here.

File: engine.py [hunk 5]
```diff
@@ -2452,7 +2543,186 @@ def _serialize_messages(self, messages: List[Dict[str, Any]]) -> str:
...
+                return cls._strip_structured_text_part(content)
+            return content if cls._structured_part_has_visible_assistant_content(content) else None
+        return content if str(content).strip() else None
+
+    @classmethod
+    def _clean_active_assistant_message(cls, msg: Dict[str, Any]) -> Dict[str, Any] | None:
+        if msg.get("role") != "assistant":
+            return msg
+        if msg.get("tool_calls"):
+            return msg
```
[the remark this answers is not in the document]
Comment (Tosko4) [reply engine.py]: Fixed in b3e36f0. Active-context cleanup now sanitizes assistant messages with tool calls too: internal-only content is cleared while the tool_call sequence stays intact. Assistant tool-call messages without a content field are left unchanged. Added coverage for tool-call cleanup and restart/rebind reconciliation.

**Comment by Tosko4** [comment]
@stephenschoettler pushed `b3e36f0` for the latest Codex P1.

Active-context cleanup now also sanitizes assistant messages that carry tool calls. Internal-only content is cleared, but the tool call sequence itself is preserved. If a tool-call assistant message has no content field, it is left alone.

I added coverage for the tool-call cleanup and restart/rebind reconciliation path. Local validation passed again, including the guardrail class, core/engine/packaging slice, full pytest, compile checks, import script py_compile, shell syntax checks, and diff check. CI is green on `b3e36f0`.

**Comment by Tosko4** [comment]
@codex review

File: engine.py [hunk 6]
```diff
@@ -2452,7 +2550,188 @@ def _serialize_messages(self, messages: List[Dict[str, Any]]) -> str:
...
+        cleaned_content = cls._sanitize_active_assistant_content(msg.get("content"))
+        if cleaned_content is None:
+            if not msg.get("tool_calls"):
+                return None
+            cleaned_content = ""
+        if cleaned_content == msg.get("content"):
+            return msg
+        cleaned = dict(msg)
+        cleaned["content"] = cleaned_content
+        return cleaned
```
[the remark this answers is not in the document]
Comment (Tosko4) [reply engine.py:2704]: Fixed in 80ef0bf. Source-id mapping now uses the active-cleanup equivalence path when matching sanitized active-context assistant turns back to the lossless stored rows. Added regressions for stripped mixed assistant content and stripped tool-call content so DAG provenance still points at the original stored turn.

**Comment by Tosko4** [comment]
@stephenschoettler pushed `80ef0bf` for the latest Codex P2.

Source-id mapping now uses the active-cleanup equivalence path when matching sanitized active-context assistant turns back to the lossless stored rows. That keeps DAG provenance tied to the original row even when provider replay strips internal assistant content.

I added regressions for stripped mixed assistant content and stripped tool-call content. Local validation passed again, including guardrail class, core/engine/packaging slice, full pytest, compile checks, import script py_compile, shell syntax checks, and diff check. CI is green on `80ef0bf`.

**Comment by Tosko4** [comment]
@codex review

**Review by stephenschoettler — COMMENTED** [review]
I did a focused maintainer pass on PR #149 at `1c60c9960690`.

Verdict:
Read-only maintainer pass completed on PR #149 at head 1c60c9960690d1c362f2727748e916b8957fabf0.

No blocking correctness, data-loss, security, performance, or maintainability regressions found in this pass.

No blockers from this pass. Safe next step: approve after a fresh GitHub state recheck.

Critical:
None.

Warnings:
None.

Suggestions:
- Optional defense-in-depth: add a future regression for stripped-but-not-dropped assistant content in a no-system/minimal transcript where a fresh delta repeats the sanitized visible suffix. The current head covers the active Codex P1 collapsed-tail case; this would harden an adjacent ambiguous restart shape.
- Optional cleanup: `_is_active_context_droppable_identity()` appears unused after the current reconciliation refactor.

Looks good:
- Active-context cleanup is correctly scoped to provider replay. Stored rows stay lossless.
- Blank structured assistant content and thinking/internal-only assistant turns are removed from active context.
- Mixed assistant content is sanitized rather than dropped: visible content survives, internal parts are stripped.
- Assistant tool-call messages are preserved; internal-only tool-call content is cleared to `""` without removing the tool call sequence.
- Existing tool-pair sanitizer still runs after cleanup, preserving provider-valid assistant/tool adjacency.
- No-compaction path avoids synthetic missing-result stubs, preventing untracked tool rows from being replayed and later duplicated.
- Restart/rebind reconciliation now mirrors active cleanup for durable-tail comparison, including dropped rows and stripped assistant content.
- Latest head includes the fix for the final Codex P1: collapsed sanitized tails no longer cause repeated visible suffix deltas to be silently skipped.
- Source-id mapping now matches sanitized active assistant turns back to original stored rows, preserving DAG provenance for stripped content.

Validation:
Clean temporary checkout:
- `/tmp/hermes-lcm-pr149-review.1Vo8n9/repo`

Commands run:
- `git diff --check origin/main...HEAD` -> pass
- `pytest -q tests/test_lcm_engine.py::TestAssemblyToolPairGuardrail` -> 25 passed
- `pytest tests/test_lcm_core.py tests/test_lcm_engine.py tests/test_packaging_install.py -q` -> 535 passed
- `pytest -q` -> 582 passed
- `python -m compileall -q .` -> pass
- `python -m py_compile scripts/import_lossless_claw.py` -> pass
- `bash -n scripts/install.sh scripts/update.sh` -> pass
- final `git status --short --branch` in scratch checkout -> clean branch `pr-149`

Only warnings observed were existing Python 3.14 deprecation warnings from the test import harness.

**Comment by Tosko4** [comment]
@codex review

**Comment by Tosko4** [comment]
@stephenschoettler pushed `7138eba` for the last replay-reconciliation edge cases.

The cleanup-sensitive raw scaffold replay path is now only treated as replay when it covers the configured fresh tail. That preserves repeated visible suffix deltas, including literal JSON assistant strings that look like structured thinking content, while still letting compacted active-context replay advance when it has fresh-tail coverage plus trailing new content.

Added regressions for:
- literal JSON assistant content on full rebind
- literal JSON assistant content after compacted rebind plus a new user row
- cleanup-sensitive scaffolded repeated tails with and without follow-up

Validation passed:
- `python -m pytest tests/test_lcm_engine.py::TestEngineABC tests/test_lcm_engine.py::TestAssemblyToolPairGuardrail -q` -> 63 passed
- `python -m pytest tests/test_lcm_core.py tests/test_lcm_engine.py tests/test_packaging_install.py -q` -> 539 passed
- `python -m pytest -q` -> 586 passed
- `python -m compileall -q .`
- `python -m py_compile scripts/import_lossless_claw.py`
- `bash -n scripts/install.sh scripts/update.sh`
- `git diff --check upstream/main...HEAD`

CI is green on `7138eba`, and the latest Codex run reported no major issues. Could you review this head?

**Review by stephenschoettler — APPROVED** [review]
I did a focused maintainer pass on PR #149 at `7138eba38d4c`.

Verdict:
No blockers from this pass. Safe next step: approve after a fresh GitHub state recheck.

Critical:
None.

Warnings:
Non-blocking cap edge:
- engine.py selects the tail under max_assembly_tokens/reserve_tokens_floor before active assistant cleanup runs.
- Reproduced in the temporary checkout: with max_assembly_tokens=80, system+visible pair=25 tokens, latest droppable thinking-only assistant=1265 tokens, _assemble_context returned only the system message.
- This is not storage/data-loss, max_assembly_tokens defaults to disabled, and it does not violate the core #140 acceptance criteria. Worth a follow-up regression if capped assembly quality matters.

Suggestions:
- Add a regression for capped assembly where the newest tail item is droppable thinking-only content and older visible tail would fit after cleanup.
- Consider pruning unused new helpers if they remain unused: _is_active_context_droppable_identity and _assistant_message_has_visible_content.

Looks good:
- Active-context cleanup removes blank structured assistant turns and thinking/internal-only assistant turns without mutating stored rows.
- Mixed assistant content is stripped rather than dropped when visible content remains.
- Assistant tool-call messages preserve tool_calls while clearing internal-only content.
- No-compaction path avoids returning synthetic untracked tool stubs.
- Restart/rebind reconciliation handles sanitized active replay, stripped assistant rows, literal JSON assistant strings, repeated visible suffix deltas, and source-id mapping back to lossless rows.
- Tests cover the prior review failures: cursor reset, two-engine rebind, source provenance, tool-call cleanup, literal JSON, and collapsed-tail replay ambiguity.

Validation:
Read-only local checkout at PR head 7138eba38d4c3fc0acbabef9ab802a17b386ff88.

Commands/results:
- gh pr view / gh pr checks: head 7138eba38d4c3fc0acbabef9ab802a17b386ff88, REVIEW_REQUIRED, BLOCKED, MERGEABLE, all visible checks SUCCESS.
- gh api reviews/comments/review-comments: inspected 18 reviews, 22 issue comments, 14 review comments.
- git diff --stat origin/main...HEAD: engine.py and tests/test_lcm_engine.py only, +1035 -27.
- python -m pytest tests/test_lcm_core.py tests/test_lcm_engine.py tests/test_packaging_install.py -q: 539 passed.
- python -m pytest -q: 586 passed.
- python -m compileall -q . && python -m py_compile scripts/import_lossless_claw.py && bash -n scripts/install.sh scripts/update.sh && git diff --check origin/main...HEAD: passed, exit 0.
- Custom read-only cap repro: confirmed the non-blocking capped-tail omission noted above.

Next action:
- Approving this head. I will merge after a final state check.

**Comment by stephenschoettler** [comment]
Thanks, this is approved, checks are green, and the current policy gates pass. I am merging this now.

File: engine.py [hunk 7]
```diff
@@ -2820,12 +3115,10 @@ def _assemble_context(
...
-        # ── Tool-pair guardrail ──
-        # Regression fix: after LCM compression, the assembled active context
-        # may contain orphan tool results (call_id with no matching assistant
-        # tool_call) or assistant tool_calls with missing results. Both violate
-        # the OpenAI message format contract and cause 400 errors from providers.
-        result = self._sanitize_tool_pairs(result)
+        # ── Active-context cleanup / tool-pair guardrail ──
+        # Drop assistant turns that carry only blank/internal structured content,
+        # then ensure provider-valid tool-call/result sequencing.
+        result = self._sanitize_active_context_messages(result)
```
[the remark this answers is not in the document]
Comment (Tosko4) [reply engine.py:3133]: Follow-up fix is open in #156. The capped tail path now runs active-context cleanup before reverse selection, with regression coverage for the dropped thinking-only assistant turn case.
